pkg/gardenlet/controller/federatedseed/shootstate/shootstate_sync_control.go

```diff
@@ -0,0 +1,144 @@
+
+package shootstate
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	extensionsclientset "github.com/gardener/gardener/pkg/client/extensions/clientset/versioned"
+	extensionsinformers "github.com/gardener/gardener/pkg/client/extensions/informers/externalversions"
+	extensionslisters "github.com/gardener/gardener/pkg/client/extensions/listers/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// SyncController replicates the extensions' resources states in Shoot's ShootState resource
+type SyncController struct {
+	k8sGardenClient      kubernetes.Interface
+	seedClient           kubernetes.Interface
+	extensionsInformers  extensionsinformers.SharedInformerFactory
+	extensionsSeedClient extensionsclientset.Interface
+
+	context             context.Context
+	cancelFunc          context.CancelFunc
+	log                 *logrus.Entry
+	infrastructuresList extensionslisters.InfrastructureLister
+	workersList         extensionslisters.WorkerLister
+
+	recorder record.EventRecorder
+
+	infrastructuresHasSynced cache.InformerSynced
+	workerHasSynced          cache.InformerSynced
+
+	infrastructureExtensionQueue workqueue.RateLimitingInterface
+	workerExtensionQueue         workqueue.RateLimitingInterface
+
+	waitGroup              sync.WaitGroup
+	numberOfRunningWorkers int
+	workerCh               chan int
+	informerShutdown       chan struct{}
+}
+
+// NewSyncController creates new controller that syncs extensions states to ShootState
+func NewSyncController(parentCtx context.Context, gardenClient, seedClient kubernetes.Interface, extensionsSeedClient extensionsclientset.Interface, extensionsInformers extensionsinformers.SharedInformerFactory, config *config.GardenletConfiguration, log *logrus.Entry, recorder record.EventRecorder) *SyncController {
+	var (
+		infraInformer        = extensionsInformers.Extensions().V1alpha1().Infrastructures()
+		workerInformer       = extensionsInformers.Extensions().V1alpha1().Workers()
+		infrastructuresList  = infraInformer.Lister()
+		workersList          = workerInformer.Lister()
+		childCtx, cancelFunc = context.WithCancel(parentCtx)
+	)
+
+	controller := &SyncController{
+		k8sGardenClient:              gardenClient,
+		seedClient:                   seedClient,
+		extensionsSeedClient:         extensionsSeedClient,
+		extensionsInformers:          extensionsInformers,
+		infrastructuresList:          infrastructuresList,
+		log:                          log,
+		workerExtensionQueue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "worker-extension-controllers"),
```

Comment:
What about all the other extension resources?

In response to https://github.com/gardener/gardener/pull/1731#discussion_r361626647: This is just the current situation, but we cannot assume that this will not change in the future. From Gardener's PoV every extension object could store state, and we should watch for it.

Reply:
We require all extension resources to be present here. The two resources mentioned in the GEP are only examples as they are the only ones that **currently** have to be synced but we cannot guarantee this in the future. Gardener should not be concerned about which extensions are synced and just blindly copy all the available states (from all extension resources) to the ShootState resource (extensions which do not require state preservation just have the state value = nil).